File: backend/exporters/ExportManager.ts
export interface ExportableTask {
  url: string;
  fileName: string;
  fileSize?: number;
}

export class ExportManager {
  /**
   * Export tasks as standard Metalink 4.0 XML (RFC 5854)
   */
  public static toMetalink(tasks: ExportableTask[]): string {
    const filesXml = tasks.map((task) => `
    <file name="${this.escapeXml(task.fileName)}">
      ${task.fileSize ? `<size>${task.fileSize}</size>` : ''}
      <url priority="1">${this.escapeXml(task.url)}</url>
    </file>`).join('\n');

    return `<?xml version="1.0" encoding="UTF-8"?>
<metalink xmlns="urn:ietf:params:xml:ns:metalink">
  <generator>Vortex Downloader 0.5.7</generator>
  <published>${new Date().toISOString()}</published>
  ${filesXml}
</metalink>`;
  }

  /**
   * Export tasks as executable cURL bash script with resume flags
   */
  public static toCurlScript(tasks: ExportableTask[]): string {
    const commands = tasks.map((task) => {
      return `echo "Downloading ${task.fileName}..."\ncurl -C - -L -o "${task.fileName}" -H "User-Agent: Mozilla/5.0" "${task.url}"\n`;
    }).join('\n');

    return `#!/usr/bin/env bash
# ==============================================================================
# Vortex Downloader - Portable cURL Export Script
# Generated: ${new Date().toISOString()}
# Usage: chmod +x download.sh && ./download.sh
# ==============================================================================

set -e
mkdir -p downloads
cd downloads

${commands}

echo "All downloads completed successfully!"
`;
  }

  /**
   * Export tasks as standard aria2 input file (aria2c -i aria2.txt)
   */
  public static toAria2Input(tasks: ExportableTask[]): string {
    return tasks.map((task) => {
      return `${task.url}\n  out=${task.fileName}\n  max-connection-per-server=16\n  split=16\n`;
    }).join('\n');
  }

  /**
   * Export tasks as standard JSON
   */
  public static toJson(tasks: ExportableTask[]): string {
    return JSON.stringify({
      version: '0.5.4',
      exportedAt: new Date().toISOString(),
      tasks
    }, null, 2);
  }

  private static escapeXml(unsafe: string): string {
    return unsafe.replace(/[<>&'"]/g, (c) => {
      switch (c) {
        case '<': return '&lt;';
        case '>': return '&gt;';
        case '&': return '&amp;';
        case '\'': return '&apos;';
        case '"': return '&quot;';
        default: return c;
      }
    });
  }
}

File: backend/exporters/export_manager.py
import json
import xml.sax.saxutils as saxutils
from datetime import datetime, timezone
from typing import List, Dict, Any, Optional

class ExportManager:
    @staticmethod
    def escape_xml(s: str) -> str:
        return saxutils.escape(s, {
            '"': "&quot;",
            "'": "&apos;"
        })

    @staticmethod
    def to_metalink(tasks: List[Dict[str, Any]]) -> str:
        files_xml_list = []
        for task in tasks:
            url = task.get('url', '')
            file_name = task.get('fileName', 'download.bin')
            size = task.get('fileSize')
            size_tag = f"    <size>{size}</size>\n" if size else ""
            files_xml_list.append(
                f'    <file name="{ExportManager.escape_xml(file_name)}">\n'
                f"{size_tag}"
                f'      <url priority="1">{ExportManager.escape_xml(url)}</url>\n'
                f'    </file>'
            )
        files_xml = "\n".join(files_xml_list)
        now_iso = datetime.now(timezone.utc).isoformat()
        return f"""<?xml version="1.0" encoding="UTF-8"?>
<metalink xmlns="urn:ietf:params:xml:ns:metalink">
  <generator>Vortex Downloader 0.6.1</generator>
  <published>{now_iso}</published>
{files_xml}
</metalink>"""

    @staticmethod
    def to_curl_script(tasks: List[Dict[str, Any]]) -> str:
        commands = []
        for task in tasks:
            url = task.get('url', '')
            file_name = task.get('fileName', 'download.bin')
            safe_name = file_name.replace('"', '\\"')
            commands.append(f'echo "Downloading {safe_name}..."\ncurl -C - -L --retry 5 --retry-delay 2 -o "{safe_name}" "{url}"')

        body = "\n\n".join(commands)
        return f"""#!/usr/bin/env bash
# ==============================================================================
# Vortex Downloader - Turbo Batch Export (cURL Script)
# Generated: {datetime.now(timezone.utc).isoformat()}
# Total Tasks: {len(tasks)}
# ==============================================================================

set -e

{body}

echo "All Vortex downloads completed successfully."
"""

    @staticmethod
    def to_aria2_input(tasks: List[Dict[str, Any]]) -> str:
        blocks = []
        for task in tasks:
            url = task.get('url', '')
            file_name = task.get('fileName', 'download.bin')
            blocks.append(f"{url}\n  out={file_name}\n  split=16\n  max-connection-per-server=16\n  continue=true")
        return "\n\n".join(blocks)

    @staticmethod
    def to_json(tasks: List[Dict[str, Any]]) -> str:
        data = {
            "version": "0.6.1",
            "generator": "Vortex Downloader",
            "exportedAt": datetime.now(timezone.utc).isoformat(),
            "totalTasks": len(tasks),
            "tasks": tasks
        }
        return json.dumps(data, indent=2)
